In [1]:
import random
import numpy as np

In [2]:
%timeit [random.random() for _ in range(100_000)]

4.09 ms ± 297 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [3]:
%timeit np.random.rand(100_000)

586 μs ± 7.17 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


## Exercise 1 Conclusion                                                                                                                                                                                                                                                       
  NumPy was much faster than pure Python for generating 100,000 random numbers.
  
  - Python list comprehension: about 4.09 ms                                                                                                                 
  - NumPy: about 586 us                                                                                                                                      
  - NumPy was about 7x faster                                                                                                                                                                                            
  This happens because Python loops run in the interpreter, while NumPy uses optimized low-level code for array operations.